allennlp/training/metrics.py

```diff
@@ -288,3 +290,196 @@ def reset(self):
+    """
+    The Conll SRL metrics are based on exact span matching. This metric
+    implements span-based precision and recall metrics for a BIO tagging
+    scheme. It will produce precision, recall and F1 measures per tag, as
+    well as overall statistics. Note that the implementation of this metric
+    is not exactly the same as the perl script used to evaluate the CONLL 2005
+    data - particularly, it does not consider continuations or reference spans
+    as constituents of the original span. However, it is a close proxy, which
+    can be helpful for judging model peformance during training.
+    """
+    def __init__(self,
+                 label_vocabulary: Dict[int, str],
+                 ignore_classes: List[int] = None) -> None:
+        """
+        Parameters
+        ----------
+        label_vocabulary : Dict[int, str], required.
+            A mapping from integer class labels to string labels. This metric
+            assumes that a BIO format is used in which the labels are of the format:
+            ["B-LABEL", "I-LABEL"].
+        ignore_classes : List[int], optional.
+            Integer classes which will be ignored when computing span
+            metrics. This is helpful for instance, to avoid computing
+            metrics for "O", "B-V" and "I-V" tags in a BIO tagging scheme
+            for semantic role labelling, which are typically not included.
+        """
+        self._label_vocabulary = label_vocabulary
+        self._ignore_classes = ignore_classes or []
+
+        # These will hold per label span counts.
+        self._true_positives = defaultdict(int)  # type: Dict[str, int]
+        self._false_positives = defaultdict(int)  # type: Dict[str, int]
+        self._false_negatives = defaultdict(int)  # type: Dict[str, int]
+
+    def __call__(self,
+                 predictions: torch.Tensor,
+                 gold_labels: torch.Tensor,
+                 mask: Optional[torch.Tensor] = None):
+        """
+        Parameters
+        ----------
+        predictions : ``torch.Tensor``, required.
+            A tensor of predictions of shape (batch_size, sequence_length, num_classes).
+        gold_labels : ``torch.Tensor``, required.
+            A tensor of integer class label of shape (batch_size, sequence_length). It must be the same
+            shape as the ``predictions`` tensor without the ``num_classes`` dimension.
+        mask: ``torch.Tensor``, optional (default = None).
+            A masking tensor the same size as ``gold_labels``.
+        """
+        if mask is None:
+            mask = torch.ones(gold_labels.size())
+        # If you actually passed in Variables here instead of Tensors, this will be a huge memory
+        # leak, because it will prevent garbage collection for the computation graph.  We'll ensure
+        # that we're using tensors here first.
+        if isinstance(predictions, Variable):
+            predictions = predictions.data.cpu()
```

Comment:
I wanted to convert the tensors to lists in the loop, just to make sure nothing weird happened - perhaps this isn't necessary and we could just do this on the tensor directly. 
You do have to do this though, as the `torch.cuda` api is not exactly the same - e.g you cannot call `.tolist()` on a cuda tensor (or possibly `Variable`), I ran into this when computing the losses previously.